rust/lance/src/dataset/write/merge_insert/exec/write.rs

```diff
@@ -0,0 +1,551 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use std::sync::{Arc, Mutex};
+
+use arrow_array::{Array, RecordBatch, UInt64Array, UInt8Array};
+use arrow_schema::Schema;
+use arrow_select;
+use datafusion::common::Result as DFResult;
+use datafusion::physical_plan::metrics::{BaselineMetrics, ExecutionPlanMetricsSet, MetricsSet};
+use datafusion::{
+    execution::{SendableRecordBatchStream, TaskContext},
+    physical_plan::{
+        execution_plan::{Boundedness, EmissionType},
+        stream::RecordBatchStreamAdapter,
+        DisplayAs, ExecutionPlan, PlanProperties,
+    },
+};
+use datafusion_physical_expr::{EquivalenceProperties, Partitioning};
+use futures::{stream, StreamExt};
+use roaring::RoaringTreemap;
+
+use crate::{
+    dataset::{
+        transaction::{Operation, Transaction},
+        write::{
+            merge_insert::{
+                assign_action::Action, exec::MergeInsertMetrics, MergeInsertParams, MergeStats,
+            },
+            write_fragments_internal, WriteParams,
+        },
+    },
+    Dataset, Result,
+};
+use lance_core::ROW_ADDR;
+use lance_table::format::Fragment;
+use std::collections::BTreeMap;
+
+/// Shared state for merge insert operations to simplify lock management
+struct MergeState {
+    /// Row addresses that need to be deleted, due to a row update or delete action
+    delete_row_addrs: RoaringTreemap,
```

Comment:
So these get accumulated in memory as the operation runs? (that's fine I think, just trying to understand)

Reply:
Yes